internal/attestations/attestations.go

```diff
@@ -0,0 +1,161 @@
+	authorizationsTreeEntryName = "authorizations"
+	initialCommitMessage        = "Initial commit"
+	defaultCommitMessage        = "Update attestations"
+)
+
+var ErrAttestationsExist = errors.New("cannot initialize attestations namespace as it exists already")
+
+// InitializeNamespace creates a namespace to store attestations for
+// verification with gittuf. The ref is created with an initial, unsigned commit
+// that is unsigned.
+func InitializeNamespace(repo *git.Repository) error {
+	if ref, err := repo.Reference(plumbing.ReferenceName(Ref), true); err != nil {
+		if !errors.Is(err, plumbing.ErrReferenceNotFound) {
+			return err
+		}
+	} else if !ref.Hash().IsZero() {
+		return ErrAttestationsExist
+	}
+
+	treeHash, err := gitinterface.WriteTree(repo, nil)
+	if err != nil {
+		return err
+	}
+
+	_, err = gitinterface.Commit(repo, treeHash, Ref, initialCommitMessage, false)
+	return err
+}
+
+// Attestations tracks all the attestations in a gittuf repository.
+type Attestations struct {
+	authorizations map[string]plumbing.Hash // refPath/from-to -> blob ID
+}
+
+// LoadCurrentAttestations inspects the repository's attestations namespace and
+// loads the current attestations.
+func LoadCurrentAttestations(repo *git.Repository) (*Attestations, error) {
+	entry, _, err := rsl.GetLatestReferenceEntryForRef(repo, Ref)
+	if err != nil {
+		if !errors.Is(err, rsl.ErrRSLEntryNotFound) {
+			return nil, err
+		}
+
+		return &Attestations{}, nil
+	}
+
+	return LoadAttestationsForEntry(repo, entry)
+}
+
+// LoadAttestationsForEntry loads the repository's attestations for a particular
+// RSL entry for the attestations namespace.
+func LoadAttestationsForEntry(repo *git.Repository, entry *rsl.ReferenceEntry) (*Attestations, error) {
+	if entry.RefName != Ref {
+		return nil, rsl.ErrRSLEntryDoesNotMatchRef
+	}
+
+	attestationsCommit, err := gitinterface.GetCommit(repo, entry.TargetID)
+	if err != nil {
+		return nil, err
+	}
+
+	attestationsRootTree, err := gitinterface.GetTree(repo, attestationsCommit.TreeHash)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(attestationsRootTree.Entries) == 0 {
+		// This happens in the initial commit for the attestations namespace,
+		// with no entries in the tree at all
+		// An RSL entry isn't created by default for that entry but if one is,
+		// we don't want to panic.
+		return &Attestations{}, nil
+	}
+
+	var authorizationsTreeID plumbing.Hash
+	for _, e := range attestationsRootTree.Entries {
+		if e.Name == authorizationsTreeEntryName {
+			authorizationsTreeID = e.Hash
+		}
+	}
+
+	authorizationsTree, err := gitinterface.GetTree(repo, authorizationsTreeID)
+	if err != nil {
+		return nil, err
+	}
+
+	attestations := &Attestations{authorizations: map[string]plumbing.Hash{}}
+
+	filesIter := authorizationsTree.Files()
+	if err := filesIter.ForEach(func(f *object.File) error {
+		attestations.authorizations[f.Name] = f.Blob.Hash
+		return nil
+	}); err != nil {
+		return nil, err
+	}
```

Comment:
You're right, this is expensive. I hadn't peeked enough but this reads the blobs in questions. I was expressly trying to avoid this and just get the filenames by reading the trees rather than the blobs as well, reading the blob in question when the specific authorization is needed. Let me see how I can rework this.